test/e2e/upgrade_kubernetes_test.go

```diff
@@ -0,0 +1,263 @@
+		listMetal3Machines(ctx, clusterClient, client.InNamespace(namespace))
+		listMachines(ctx, clusterClient, client.InNamespace(namespace))
+		listNodes(ctx, targetClusterClient)
+
+		// Download node image
+		By("Download image")
+		osType := strings.ToLower(os.Getenv("OS"))
+		Expect(osType).ToNot(Equal(""))
+		imageNamePrefix = "UBUNTU_22.04_NODE_IMAGE_K8S"
+		imageName := fmt.Sprintf("%s_%s.qcow2", imageNamePrefix, upgradedK8sVersion)
+		Logf("IMAGE_NAME: %v", imageName)
+		rawImageName := fmt.Sprintf("%s_%s-raw.img", imageNamePrefix, upgradedK8sVersion)
+		Logf("RAW_IMAGE_NAME: %v", rawImageName)
+		imageLocation := fmt.Sprintf("%s_%s/", artifactoryURL, upgradedK8sVersion)
+		Logf("IMAGE_LOCATION: %v", imageLocation)
+		imageURL := fmt.Sprintf("%s/%s", imagesURL, rawImageName)
+		Logf("IMAGE_URL: %v", imageURL)
+		imageChecksum := fmt.Sprintf("%s/%s.md5sum", imagesURL, rawImageName)
+		Logf("IMAGE_CHECKSUM: %v", imageChecksum)
+
+		// Check if node image with upgraded k8s version exist, if not download it
+		if _, err := os.Stat(fmt.Sprintf("%s/%s", ironicImageDir, rawImageName)); err == nil {
+			Logf("Local image %v is found", rawImageName)
+		} else if os.IsNotExist(err) {
+			Logf("Local image %v/%v is not found", ironicImageDir, rawImageName)
+			err = downloadFile(fmt.Sprintf("%s/%s", ironicImageDir, imageName), fmt.Sprintf("%s/%s", imageLocation, imageName))
+			Expect(err).To(BeNil())
+			cmd := exec.Command("qemu-img", "convert", "-O", "raw", fmt.Sprintf("%s/%s", ironicImageDir, imageName), fmt.Sprintf("%s/%s", ironicImageDir, rawImageName))
+			err = cmd.Run()
+			Expect(err).To(BeNil())
+			cmd = exec.Command("md5sum", fmt.Sprintf("%s/%s", ironicImageDir, rawImageName))
+			output, err := cmd.CombinedOutput()
+			Expect(err).To(BeNil())
+			md5sum := strings.Fields(string(output))[0]
+			err = os.WriteFile(fmt.Sprintf("%s/%s.md5sum", ironicImageDir, rawImageName), []byte(md5sum), 0777)
+			Expect(err).To(BeNil())
+		} else {
+			fmt.Fprintf(GinkgoWriter, "ERROR: %v\n", err)
+			os.Exit(1)
+		}
+
+		By("Update KCP Metal3MachineTemplate with upgraded image to boot")
+		m3machineTemplateName := fmt.Sprintf("%s-controlplane", clusterName)
+		updateBootImage(m3machineTemplateName, clusterClient, imageURL, imageChecksum, "raw", "md5")
+
+		Byf("Update KCP to upgrade k8s version and binaries from %s to %s", kubernetesVersion, upgradedK8sVersion)
+		kcpObj := framework.GetKubeadmControlPlaneByCluster(ctx, framework.GetKubeadmControlPlaneByClusterInput{
+			Lister:      clusterClient,
+			ClusterName: clusterName,
+			Namespace:   namespace,
+		})
+		patch := []byte(fmt.Sprintf(`{
+			"spec": {
+				"rolloutStrategy": {
+					"rollingUpdate": {
+						"maxSurge": 0
+					}
+				},
+				"version": "%s"
+			}
+		}`, upgradedK8sVersion))
+		err := clusterClient.Patch(ctx, kcpObj, client.RawPatch(types.MergePatchType, patch))
+		Expect(err).To(BeNil(), "Failed to patch KubeadmControlPlane")
```

Comment:
I'm a bit afraid of this actually. It is the same as in node reuse but I was not aware of it. In the ansible tests we create the new templates and then in [one single patch](https://github.com/metal3-io/metal3-dev-env/blob/c2ca4f453f31bd2cd7272ba62ba9ea731e86f484/tests/roles/run_tests/tasks/upgrade.yml#L378-L390) we update the infrastructureRef and version. But here we do two patches! And it even looks like we are patching the template itself. Maybe I'm missing something? I thought it was impossible to patch the template and this is why we create new templates.

My issue with 2 separate patches is that we can get inconsistencies. For a little while we will have wrong k8s version for the specified boot image.

Reply:
solved. Now it creates new m3machinetemplate